### PR TITLE
Add binary_mclp_distance_matrix into pyspatialopt.models

### PR DESCRIPTION
**Pull request by huanfachen**
A binary MCLP model using a distance matrix

### CODE REVIEW

File: examples/binary_lscp_arcpy.py [hunk 1]
```diff
@@ -17,18 +18,18 @@
     sh = logging.StreamHandler(sys.stdout)
     sh.setFormatter(formatter)
     logger.addHandler(sh)
-
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
Comment (apulverizer) [inline examples/binary_lscp_arcpy.py]: can you change this to `script_dir` to make it more "pythonic"?

I assume this was to run it from any directory?

File: examples/binary_mclp_arcpy.py [hunk 2]
```diff
@@ -17,14 +17,19 @@
     sh = logging.StreamHandler(sys.stdout)
     sh.setFormatter(formatter)
     logger.addHandler(sh)
-
+    #import os
```
Comment (apulverizer) [inline examples/binary_mclp_arcpy.py]: can you remove these commented out lines?

File: examples/binary_mclp_arcpy.py [hunk 3]
```diff
@@ -17,14 +17,19 @@
...
+    #print(os.getcwd())
+    import os
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
Comment (apulverizer) [inline examples/binary_mclp_arcpy.py]: change to `script_dir`

File: examples/binary_mclp_distance_matrix_example.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    logger = logging.getLogger()
+    logger.setLevel(logging.INFO)
+    formatter = formatter = logging.Formatter('%(asctime)s %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p')
+    # setup stream handler to console output
+    sh = logging.StreamHandler(sys.stdout)
+    sh.setFormatter(formatter)
+    logger.addHandler(sh)
+
+    import os
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
Comment (apulverizer) [inline examples/binary_mclp_distance_matrix_example.py]: change to `script_dir`

**Review by apulverizer — CHANGES_REQUESTED** [review]
Thanks for the PR, I haven't actually run it but it looks pretty good. 

Can you clean up the code a bit by removing any commented out code?
There were a couple of other things noted in my comments.

I'm not sure what the Jupyter notebook was trying to show. That would need to be cleaned up a lot if you want that merged, otherwise can you remove that from the PR?

File: examples/binary_lscp_arcpy.py (hunk 1 above)
Comment (huanfachen) [reply examples/binary_lscp_arcpy.py]: Hi. I dont want to commit the changes in the examples/binary_lscp_arcpy.py and binary_mclp_arcpy.py files so I put them in the .gitignore files. However, the .gitignore file does not work here and these changes were submitted. 
Any idea?

File: .gitignore [hunk 5]
```diff
@@ -69,3 +69,13 @@ target/
 .xml
 .shx
 .cpg
+
+#IDE folder
+.vscode
```
Comment (apulverizer) [inline .gitignore]: keep this line but remove all the others

File: examples/binary_lscp_arcpy.py (hunk 1 above)
Comment (apulverizer) [reply examples/binary_lscp_arcpy.py]: Can you make a copy of those files and re-name it something else? Then you can have those file locally for yourself.

Then just copy the original files on top of the ones you modified. Then commit those 2 files. That should make those files unchanged.

Or you can try and revert your last couple commits for those specific files if you're familiar with git.

File: .gitignore (hunk 5 above)
Comment (huanfachen) [reply .gitignore]: I donot get it. If so (all the other lines are removed), the IPython notebook file will be committed, which is not my intention.

Comment (apulverizer) [reply .gitignore]: If I understand correctly, you committed a bunch of files and pushed those changes. When you got to commit the changes you should only add/commit the specific files that you want to add to the repo or update in the repo.

So say I add some new files locally called `test.py` and `test2.py`. And I only want to commit `test.py` then I would just do `git add test.py` and then `git commit -m "add test.py"`  and then git push. Then only `test.py` would be committed and pushed the server. And I don't have to update the `.gitignore` file. 

I'm guessing maybe you committed an entire directory and that included all of the files that were add/changed.

Comment (huanfachen) [reply .gitignore]: Thanks, now I understand. I figure out how to use git to revert files and remove cache. Please have a look at my latest commit. Let me know if there is anything that needs to be improved.

File: pyspatialopt/models/binary_mclp_distance_matrix.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
+# -*- coding: UTF-8 -*-
+import logging
+import sys
```
Comment (apulverizer) [inline pyspatialopt/models/binary_mclp_distance_matrix.py]: I pulled down your branch and the `sys` and `arcpy` modules are not used, can you remove them?

File: pyspatialopt/models/binary_mclp_distance_matrix.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+import arcpy
+import pulp
+import csv
+import sys
+import os
+from pyspatialopt.models import utilities
+from pyspatialopt.models import covering
+
```
Comment (apulverizer) [inline pyspatialopt/models/binary_mclp_distance_matrix.py]: Add an extra new line to follow [Pep8](https://www.python.org/dev/peps/pep-0008/) styling guidelines

File: tests/binary_mclp_distance_matrix_test.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        workspace_path = r"../sample_data"
+        file_distance_matrix = "service_area_demand_point_distance_matrix.csv"
+        service_dist = 5000
+        # dict_num_facility_percent_coverage = {1:5.0/15*100, 2:9.0/15*100, 3:12.0/15*100, 4:14.0/15*100, 5:15.0/15*100}
+        dict_num_facility_percent_coverage = {5:52.797393302}
+        for num_facility, percent_coverage in dict_num_facility_percent_coverage.items():
+            res_coverag = binary_mclp_distance_matrix.binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, workspace_path=workspace_path)
+            print("A coverage of {0} is obtained with {1} facilities".format(res_coverag["percent_demand_coverage"], num_facility))
+            self.assertAlmostEquals(res_coverag["percent_demand_coverage"], percent_coverage, places=5)
+
```
Comment (apulverizer) [inline tests/binary_mclp_distance_matrix_test.py]: Add another new line here for PEP 8

File: pyspatialopt/models/binary_mclp_distance_matrix.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                output["demand"][str(row[dl_id_field])]["demand"]
+            output["demand"][str(row[dl_id_field])]["coverage"][fl_variable_name][str(row[fl_id_field])] = 1
+
+    # summary
+    for row in output["demand"].values():
+        output["totalServiceableDemand"] += row["serviceableDemand"]
+        output["totalDemand"] += row["demand"]    
+    logging.getLogger().info("Binary coverage successfully generated.")
+    return output
+
```
Comment (apulverizer) [inline pyspatialopt/models/binary_mclp_distance_matrix.py]: Add a new line here for Pep 8

File: pyspatialopt/models/binary_mclp_distance_matrix.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    
+    result_coverage = {"number_facility": num_facility,
+    "number_facility_chosen": len(set_facility_id_chosen),
+    "set_facility_id_chosen": set_facility_id_chosen,
+    "total_demand": dict_coverage["totalDemand"], 
+    "percent_demand_coverage":(100 * total_demand_covered) / dict_coverage["totalDemand"]
+    }
+
+    return result_coverage
+
```
Comment (apulverizer) [inline pyspatialopt/models/binary_mclp_distance_matrix.py]: Delete these extra ilnes here. There should be one newline after your `return` statement per PEP 8.

File: pyspatialopt/models/binary_mclp_distance_matrix.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            output["demand"][str(row[dl_id_field])]["coverage"][fl_variable_name][str(row[fl_id_field])] = 1
+
+    # summary
+    for row in output["demand"].values():
+        output["totalServiceableDemand"] += row["serviceableDemand"]
+        output["totalDemand"] += row["demand"]    
+    logging.getLogger().info("Binary coverage successfully generated.")
+    return output
+
+def binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, list_field_req = ["facility_id", "demand_id", "demand", "distance"], facility_variable_name = "facility", workspace_path = "."):
```
Comment (apulverizer) [inline pyspatialopt/models/binary_mclp_distance_matrix.py]: I didn't see this before. This is one of those python gotchas. A list is immutable and should never be used as a default value when it's a parameter. You can read more about it [here](https://docs.python-guide.org/writing/gotchas/)

Instead set `list_field_req=None` and then do a check in the function like:
```python
if list_field_req is None:
    list_field_req = ["facility_id", "demand_id", "demand", "distance"]

**Review by apulverizer — CHANGES_REQUESTED** [review]
Thanks for making the changes. I took another look and noticed a few style changes that should be made. In addition to the specific comments, in the functions when you set default arguments like:
```python
def some_function(facility_variable_name = "facility", workspace_path = ".")
```
there should not be any spaces around the `=` signs. It should look like:
```python
def some_function(facility_variable_name="facility", workspace_path=".")
```

Otherwise it looks good and I was able to run the test you provided (after I commented out the `import arcpy` line)

If you make the suggested changes I'd be happy to merge it.

File: pyspatialopt/models/binary_mclp_distance_matrix.py (hunk 10 above)
Comment (huanfachen) [reply pyspatialopt/models/binary_mclp_distance_matrix.py]: A bit confused here: in pep8 it says "Surround top-level function and class definitions with two blank lines". Looks like two blank lines are needed after the `return` statement?

File: pyspatialopt/models/binary_mclp_distance_matrix.py (hunk 11 above)
Comment (huanfachen) [reply pyspatialopt/models/binary_mclp_distance_matrix.py]: Good point. Do you mean "a list is mutable and should never be used as a default value"?

File: examples/binary_mclp_distance_matrix_example.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+            for row in csv.DictReader(csvfile, skipinitialspace=True)
+            ]
+
+    # test if it contains the required field
+    item_pairwise_distance = dict_pairwise_distance[1]
+
+    for field in list_field_req:
+        if field not in item_pairwise_distance.keys():
+            logger.info("Error: this field {} not found in the distance csv".format(field))
+            sys.exit(0)
```
Comment (apulverizer) [inline examples/binary_mclp_distance_matrix_example.py]: sys isn't imported in this file . Sorry, I may have mad you remove that earlier.

File: tests/binary_mclp_distance_matrix_test.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        workspace_path = r"../sample_data"
+        file_distance_matrix = "simple_case_distance_matrix.csv"
+        service_dist = 10
+        dict_num_facility_percent_coverage = {1: 5.0/15*100, 2: 9.0/15*100, 3: 12.0/15*100, 4: 14.0/15*100, 5: 15.0/15*100}
+        for num_facility, percent_coverage in dict_num_facility_percent_coverage.items():
+            res_coverag = binary_mclp_distance_matrix.binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, workspace_path=workspace_path)
+            print("A coverage of {0} is obtained with {1} facilities".format(res_coverag["percent_demand_coverage"], num_facility))
+            self.assertAlmostEquals(res_coverag["percent_demand_coverage"], percent_coverage)
+
+    def test_simple_case(self):
```
Comment (apulverizer) [inline tests/binary_mclp_distance_matrix_test.py]: This function name is duplicated

**Review by apulverizer — CHANGES_REQUESTED** [review]
Looks good other than those two minor things. Thanks for re-formatting it.

**Pull request merged**